### PR TITLE
[R4R]reannounce local pending transactions

### PR DESCRIPTION
**Pull request by KeefeL**
### Description

When the network traffic is heavy, transactions that have been broadcast may be truncated in peers' tx_pools, while these local pending transactions have been marked as known, the source node won't broadcast them again, so they have no chance to be mined. 

We add a timer to check whether the local pending transactions stay in tx_pool exceed a threshold, once transactions exceed the threshold, they will be announced to some peers again. The  TxPool.ReannounceTime is used to control the threshold,  the default value of TxPool.ReannounceTime is too big to be exceeded, users can config this parameter to enable this feature.

Also need to pay attention, this feature only announce the local pending transactions. If your node is configured TxPool.NoLocals=false, then the transactions sent to this RPC node will be treated as local transactions, otherwise you should add your addresses to the TxPool.Locals, only the transactions from the local addresses will be treated as local transactions.

### Rationale

This feature is introduced to handle some extreme situations, it won't affect the original function, and it is disabled by default. Users can enable this feature by setting TxPool.ReannouceTime according to their needs.

### Example
Set TxPool.ReannounceTime to 5 minutes:
geth --config ./config.toml --datadir ./node  --cache 8000 --rpc.allow-unprotected-txs --txlookuplimit 0 --txpool.reannouncetime 5m

### Changes

Notable changes: 
* add --txpool.reannouncetime parameter, default value is 10 years, minimum value is 1 minute.

### CODE REVIEW

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -398,6 +398,11 @@ var (
 		Usage: "Maximum amount of time non-executable transaction are queued",
 		Value: ethconfig.Defaults.TxPool.Lifetime,
 	}
+	TxPoolReannounceTimeFlag = cli.DurationFlag{
+		Name:  "txpool.reannouncetime",
+		Usage: "Minimum amount of time for announcing local pending transactions again",
```
Comment (guagualvcha) [inline cmd/utils/flags.go:403]: `Minimum amount of time` sound weried.

**Review by guagualvcha — APPROVED** [review]

**Review by yutianwu — APPROVED** [review]

File: core/tx_pool.go [hunk 2]
```diff
@@ -323,9 +334,10 @@ func (pool *TxPool) loop() {
 	var (
 		prevPending, prevQueued, prevStales int
 		// Start the stats reporting and transaction eviction tickers
-		report  = time.NewTicker(statsReportInterval)
-		evict   = time.NewTicker(evictionInterval)
-		journal = time.NewTicker(pool.config.Rejournal)
+		report     = time.NewTicker(statsReportInterval)
+		evict      = time.NewTicker(evictionInterval)
+		reannounce = time.NewTicker(reannounceInterval)
```
Comment (j75689) [inline core/tx_pool.go:339]: please add `defer reannounce .Stop()` below

File: core/tx_pool.go [hunk 3]
```diff
@@ -378,6 +390,33 @@ func (pool *TxPool) loop() {
 			}
 			pool.mu.Unlock()
 
+		case <-reannounce.C:
+			pool.mu.RLock()
```
[the remark this answers is not in the document]
Comment (KeefeL) [reply core/tx_pool.go:394]: I think we should use the read lock, other goroutines may write the tx_pool, and then the iteration maybe wrong.

[the remark this answers is not in the document]
Comment (j75689) [reply core/tx_pool.go:394]: but `reannoTxs` will not be executed here, it is a sub-function you send to `pool.reannoTxFeed`. I think maybe you want to lock the `reannoTxs` function

[the remark this answers is not in the document]
Comment (j75689) [reply core/tx_pool.go:394]: sorry, my bad.
I got it wrong

File: cmd/utils/flags.go (hunk 1 above)
Comment (KeefeL) [reply cmd/utils/flags.go]: done

File: core/tx_pool.go (hunk 2 above)
Comment (KeefeL) [reply core/tx_pool.go:339]: done

File: core/tx_pool.go [hunk 4]
```diff
@@ -208,6 +214,10 @@ func (config *TxPoolConfig) sanitize() TxPoolConfig {
 		log.Warn("Sanitizing invalid txpool lifetime", "provided", conf.Lifetime, "updated", DefaultTxPoolConfig.Lifetime)
 		conf.Lifetime = DefaultTxPoolConfig.Lifetime
 	}
+	if conf.ReannounceTime < time.Minute {
```
[the remark this answers is not in the document]
Comment (KeefeL) [reply core/tx_pool.go:217]: done

**Review by j75689 — APPROVED** [review]

**Pull request merged**